Image class functionality to be reviewed

j-tr opened the pull request.

j-tr committed:
first draft of Bitmap class

j-tr committed:
Image class

j-tr committed:
working image class

Lichtso commented on source/llassetgen-cmd/main.cpp:
```diff
@@ -5,14 +5,68 @@
     llassetgen::init();
 
-    std::unique_ptr<DistanceTransform> dt(new DeadReckoning());
+	// creating the face
```
Please move this to:
source/tests/llassetgen-tests/main.cpp

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,293 @@
+		max_y = _max_y;
+		bit_depth = _bit_depth;
+	}
+
```
Image::at and void Image::put are overcomplicated and inefficient (e.g. a / c + b / c = (a+b) / c)
also the idea of stride is not to be a alias for width but to include bitDepth.
Furthermore, this implementation contains a loop and is not endianness aware.
And they might be renamed to get / set instead of at / put.

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,293 @@
+		width = _width;
+		height = _height;
+		bit_depth = _bit_depth;
+		stride = _width;
```
stride = _width * (_bitDepth / 8);

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,293 @@
+	void Image::exportPng(const std::string filepath) {
+		std::ofstream out_file(filepath, std::ofstream::out | std::ofstream::binary);
+		if (!out_file.good()) {
+			std::cerr << "couldnt open file";
```
Couldn't / Could not
and printing the filepath would be nice as well

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,293 @@
+		png_destroy_read_struct(&png, &info, (png_infopp)0);
+		in_file.close();
+
+		data = std::shared_ptr<uint8_t>(new uint8_t[width * height * bit_depth / 8]);
```
stride * height

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,293 @@
+		png_read_update_info(png, info);
+		uint8_t channels = png_get_channels(png, info);
+		
+		uint8_t stride = width * bit_depth * channels / 8;
```
stride in this local scope here shadows the member / instance variable

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,293 @@
+		data = std::shared_ptr<uint8_t>(new uint8_t[width * height * bit_depth / 8]);
+		for (size_t y = 0; y < height; y++) {
+			for (size_t x = 0; x < width; x++) {
+					data.get()[y * width * bit_depth / 8 + x * bit_depth / 8] = multi_channel_data.get()[y * stride + x * bit_depth / 8 * channels];
```
Why not use Image::put here?

j-tr committed:
improved regarding some of alex comments

scheibel commented on source/llassetgen-cmd/main.cpp:
```diff
@@ -5,14 +5,68 @@
+	FT_Face face;
+	std::string font_file = "C:/Windows/Fonts/Arial.ttf"; // path to local font
+	char letter = 'A';
+	assert(FT_New_Face(freetype, font_file.c_str(), 0, &face) == 0);
```
Beware that this code isn't executed for release executables!
I presume you want to execute this code and check the return value as assertion when in debug mode.

scheibel commented on source/llassetgen-cmd/main.cpp:
```diff
@@ -5,14 +5,68 @@
+	blank_32bit.put(4, 5, 1234);
+
+	// get pixels and test
+	for (int i = 0; i < blank_32bit.get_height(); i++) {
```
Rather cumbersome code to clear the whole image and set two specific pixels.

scheibel commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,39 @@
+			LLASSETGEN_API size_t get_bit_depth();
+			LLASSETGEN_API Image view(const size_t _min_x, const size_t _max_x, const size_t _min_y, const size_t _max_y);
+		private:
+			Image::Image(const size_t _min_x, const size_t _max_x, const size_t _min_y, const size_t _max_y, const uint8_t _bit_depth, const size_t _stride, const std::shared_ptr<uint8_t> _data);
```
Unnecessary `const` in declaration (however, for the definition the `const` is fine).

scheibel commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,39 @@
+	class Image {
+		public:
+			LLASSETGEN_API Image(const FT_Bitmap &ft_bitmap);
+			LLASSETGEN_API Image(const std::string filepath);
```
Parameter passing by copy? Why not by reference?

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,294 @@
+			size_t bit_pos = offset_x % (8 / bit_depth);
+			byte <<= bit_pos * bit_depth;
+			byte >>= 8 - bit_depth;
+			return (uint32_t)byte;
```
There is a C++ `static_cast<std::uint32_t>()` for that.

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,294 @@
+			in_byte <<= 8 - bit_pos * bit_depth - bit_depth;
+			mask <<= 8 - bit_pos * bit_depth - bit_depth;
+
+			data.get()[offset_y * stride + offset_x * bit_depth / 8] = data.get()[offset_y * stride + offset_x * bit_depth / 8] & ~mask | in_byte;
```
Unnecessary recompution of index (`offset_y * stride + offset_x * bit_depth / 8`).

j-tr committed:
monochrome bug fixed

j-tr committed:
load freetype bitmap in existing image

j-tr committed:
templated image class

Lichtso commented on source/llassetgen/source/Image.cpp:
```diff
@@ -184,40 +195,63 @@ namespace llassetgen {
+			// scale down to 16 bit int grayscale
+			
+			if (max > pow(2, bit_depth)) {
+				max = pow(2, bit_depth);
```
pow(2, bit_depth) is using floating point arithmetics
Use bit shifting instead: 1 << bit_depth

j-tr committed:
removed useless pow

Lichtso commented:
Could you move your tests in main.cpp to the designated test file?
I can then approve your PR

Lichtso committed:
Reverted changes in CLI/main.cpp to solve merge conflicts

Lichtso committed:
Merge branch 'master' into bitmap_class

Lichtso committed:
Silenced compiler warnings

Lichtso committed:
Copied include fix for travis from DistanceTransform.cpp

Lichtso committed:
Fixed include chaos

Lichtso committed:
Using Vec2 in Image class

Lichtso committed:
Code cleanup and added missing clamp() in Image::exportPng()

Lichtso committed:
Fixed duplicate symbols

Lichtso committed:
Renamed at/put => getPixel/setPixel

Lichtso committed:
Started porting DistanceTransform to use Image

Lichtso committed:
Ported DeadReckoning

Lichtso committed:
Ported ParabolaEnvelope

Lichtso committed:
Added example usage of ported DistanceTransform

scheibel commented on source/llassetgen/include/llassetgen/DistanceTransform.h:
```diff
@@ -1,46 +1,24 @@
-        LLASSETGEN_API void importPng(std::string path);
-        LLASSETGEN_API void exportPng(std::string path, OutputType blackDistance, OutputType whiteDistance,
-                                      DimensionType bitDepth);
+        const Image &input, &output;
```
Please split bulk declaration into separate lines.

scheibel commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,40 @@
+namespace llassetgen {
+	class Image {
+		public:
+			LLASSETGEN_API Image(const std::string &filepath);
```
Please use 4 spaces instead of tabs.

scheibel commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,40 @@
+struct FT_Bitmap_;
+struct png_struct_def;
+
+namespace llassetgen {
```
We don't indent within a namespace and place the opening brace on the next line.

anne-gropler replied:
In this project we use a formatting tool which places the braces like that.

scheibel commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,162 @@
+        if (output.is_valid(target) && getPixel<OutputType>(target) + distance < getPixel<OutputType>(pos)) {
             posAt(pos) = target = posAt(target);
-            outputAt(pos) = std::sqrt(square(pos.x - target.x) + square(pos.y - target.y));
+            setPixel<OutputType>(pos, std::sqrt(square(pos.x - target.x) + square(pos.y - target.y)));
```
I think glm [1] would be a valid dependency for this project.

[1] https://github.com/g-truc/glm

Lichtso replied:
I tried using alternatives like `std::hypot(x, y)` but it is not that easy because of the signed / unsigned types

anne-gropler replied:
In llassetgen-rendering I use glm, too. So if we need glm as a main dependency, I would benefit from it, too.

@Lichtso I think @scheibel meant to use glm instead of implementing vector maths by ourselfs.

scheibel commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,162 @@
-                    outputAt(offset+i*pitch) = square((i < (length+begin)/2) ? i-begin+1 : length-i);
-            }
+            for(DimensionType i = begin; i < length; ++i)
+                setPixel<OutputType, flipped>({i, offset}, (begin == 0)
```
This could profit from a comment.

Lichtso replied:
Done

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+		max.y = height;
+        bit_depth = _bit_depth;
+        stride = (size_t)std::ceil(float(get_width() * (float(bit_depth) / 8.0f)));
+		data = std::shared_ptr<uint8_t>(new uint8_t[get_height() * stride]());
```
I suggest using `std::make_shared`.

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+			return *reinterpret_cast<pixelType*>(&casted);
+		} else {
+			uint32_t return_data = 0;
+			for (size_t byte_pos = 0; byte_pos < bit_depth / 8; ++byte_pos) {
```
This code looks optimized, but I presume it misses some optimization potential.

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+        data = _data;
+	}
+
+    bool Image::is_valid(const Vec2<size_t> pos) const {
```
Take a const reference as parameter?

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+			info,
+			get_width(),
+			get_height(),
+			(bit_depth < 16) ? bit_depth : 16,
```
This looks like a `std::min`.

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+
+	void Image::read_data(png_structp png, png_bytep data, png_size_t length) {
+		png_voidp a = png_get_io_ptr(png);
+		((std::istream*)a)->read((char*)data, length);
```
I suggest to add an assertion that a `dynamic_cast` to `std::istream*` results in a valid pointer.

scheibel commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+		stride = get_width() * (bit_depth / 8);
+
+		std::unique_ptr<png_bytep> row_ptrs(new png_bytep[get_height()]);
+		std::unique_ptr<uint8_t> multi_channel_data(new uint8_t[get_height() * stride * channels]);
```
What is the advantage of this code over a `std::vector<uint8_t>`?

Lichtso committed:
Cleaned up source code
Added comments
Fixed a small bug in ParabolaEnvelope

Lichtso committed:
Reactivated CXX_VISIBILITY_PRESET "hidden"
Added LLASSETGEN_API / LLASSETGEN_NO_EXPORT in interfaces

j-tr committed:
moved tests to the right place

anne-gropler commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,358 @@
+		data = std::shared_ptr<uint8_t>(new uint8_t[get_height() * stride]());
+	}
+
+	size_t Image::get_width() const {
```
I am not sure if we agreed on under_score for variable names. I thought we use camelCase?

get/set: if it's a getter, just name it width(). At least that's a naming convention that I follow a lot, but it's optional. We might discuss that in our project group, if you want.

j-tr committed:
reduce bit depth while loading

anne-gropler committed:
fix uninitialized freetype library

Lichtso committed:
Tabs to spaces

Lichtso committed:
Cleaned up source code to fix warnings

Lichtso committed:
Added DistanceTransformTest

Lichtso committed:
Cleaned up source code

j-tr committed:
fixed big pngs

j-tr committed:
fix freetype release init

Blaidd-Drwg commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,162 @@
-        return (input[offset / 8] >> (offset % 8)) & 1;
+    template<typename PixelType, bool flipped, bool invalidBounds>
+    PixelType DistanceTransform::getPixel(PositionType pos) {
+        const Image& image = std::is_same<PixelType, InputType>::value ? input : output;
```
Is it idiomatic to return different data depending on the template type (in this case, pixels from different images)? I found it pretty confusing at first.

Lichtso replied:
The alternative is to use a even less obvious types like bool to switch between them.
Thus, I thought reading InputType / OutputType is better than just true / false.

Blaidd-Drwg replied:
It should at least be documented. Also, it isn't needed in `setPixel` since AFAIK we never modify the input, only the output.

Blaidd-Drwg replied:
An alternative would be having two functions `getInputPixel` and `getOutputPixel` which pass the correct image as an argument to `getPixel`, though I don't know if the additional function call would affect performance.

Lichtso replied:
I combined them to avoid code redundancy

Blaidd-Drwg replied:
You'd be adding a one-liner for `getInputPixel` and a two-liner for `getOutputPixel`. As far as redundancy is concerned, I think it's a negligible trade-off for the clearer interface.

j-tr committed:
change testfile path

Blaidd-Drwg commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,367 @@
+        Vec2<size_t> offset = min + pos;
+        if (bitDepth <= 8) {
+            uint8_t mask = 0xFF;
+            uint8_t in_byte = static_cast<uint8_t>(*reinterpret_cast<uint32_t*>(&in));
```
In [DistanceTransform.cpp#L55](https://github.com/hpicgs/openll-asset-generator/blob/bitmap_class/source/llassetgen/source/DistanceTransform.cpp#L55) this function is called either with `in` set to 0 or infinity. In both cases `in_byte` would be 0.

Blaidd-Drwg committed:
Rechange testfile path

Blaidd-Drwg committed:
Merge branch 'master' into bitmap_class

Blaidd-Drwg committed:
Fix bug that read every bitmap as entirely white

Blaidd-Drwg commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,166 @@
+    PixelType DistanceTransform::getPixel(PositionType pos) {
+        const Image& image = std::is_same<PixelType, InputType>::value ? input : output;
+        if(invalidBounds && !image.isValid(pos))
+            return 0;
```
This makes every pixel outside of the bounds black in the eyes of the distance transform. Since we render black characters on white background, distance fields have a white border around them.
![testfontatlas_rgb](https://user-images.githubusercontent.com/6051384/37316772-4f68d9f2-2661-11e8-80fc-30e1b0935025.png)

Blaidd-Drwg replied:
Changing the value to 1 fixes the problem when DeadReckoning is used, though I don't know if this could break `getPixel<OutputType>`. ParabolaEnvelope creates a white border regardless of the value.

Lichtso replied:
The reason for the `TODO` at the end of the line
`(value >= std::numeric_limits<InputType>::max()/2) : value; // TODO`
was that this check should not happen in DistanceTransform but in Image.
The bit depth reduction was not implemented back then, thus this ugly solution.

Lichtso replied:
So we have to decide what black / white interpretation we want and use a consistent one and remove this discretization here.

Blaidd-Drwg replied:
I see. Could you move the check to Image then?

Blaidd-Drwg replied:
Assuming we choose the current b/w interpretation, do you have an idea how to fix ParabolaEnvelope?

Lichtso replied:
Doesn't TTF use 0 / black for empty and 1 / white for set?
Why would you invert the input, it only costs computation power and time...
Please keep in mind that the output can simply be inverted independent of the input by flipping black / white and min / max parameters of Image::exportPng().

Lichtso committed:
DistanceTransform now uses bit depth reduction of Image class

Lichtso committed:
Removed color inversion at TTF loading
Refactored and cleaned up source code

Felerius committed:
Fix includes on windows

Lichtso committed:
Added cleared padding option to view
Added Image::fillRect() method
Fixed views of views missing translation
Fixed Image::load()
Removed pointer de/referencing from reinterpret_casts

Blaidd-Drwg commented on source/llassetgen/source/Image.cpp:
```diff
@@ -36,8 +36,16 @@ namespace llassetgen {
+            fillRect(outerMin, Vec2<size_t>(innerMax.x, innerMin.y));
+            fillRect(Vec2<size_t>(innerMax.x, outerMin.y), Vec2<size_t>(outerMax.x, innerMax.y));
+            fillRect(Vec2<size_t>(innerMin.x, innerMax.y), outerMax);
+            fillRect(Vec2<size_t>(outerMin.x, innerMin.y), Vec2<size_t>(innerMin.x, outerMax.y));
```
Can be made more concise using initializer lists.

Blaidd-Drwg commented on source/llassetgen/source/Image.cpp:
```diff
@@ -101,17 +109,43 @@ namespace llassetgen {
+    void Image::fillRect(Vec2<size_t> _min, Vec2<size_t> _max, pixelType in) const {
+        for(size_t y = _min.y; y < _max.y; y++) {
+            for(size_t x = _min.x; x < _max.x; x++) {
+                setPixel<uint16_t>({x, y}, in);
```
Why `uint16_t`?

Blaidd-Drwg commented on source/llassetgen/source/Image.cpp:
```diff
@@ -36,8 +36,16 @@ namespace llassetgen {
-        return Image(_min, _max, stride, bitDepth, data);
+    Image Image::view(Vec2<size_t> outerMin, Vec2<size_t> outerMax, size_t padding) {
+        assert(outerMax.x <= getWidth() && outerMax.y <= getHeight());
+        Vec2<size_t> vec(padding, padding), innerMin = min + outerMin + vec, innerMax = min + outerMax - vec;
```
Could you put these statements on separate lines and give `vec` a more descriptive name like `paddingVec`?

Blaidd-Drwg replied:
Why are `min` and `max` added to `innerMin` and `innerMax` but not to `outerMin` and `outerMax`?

Lichtso committed:
Fixed some minor issues

Blaidd-Drwg committed:
clang-format

Blaidd-Drwg committed:
Fix some clang-format uglification

Blaidd-Drwg committed:
Change test font file

Blaidd-Drwg committed:
Add font license

Blaidd-Drwg committed:
Done a whoopsie on the license there

Lichtso committed:
Replaced padding clearing by separate method

Lichtso reviewed (approved):
The most important features are implemented and tests running. Further optimizations will be postponed to another branch.

Felerius commented on source/llassetgen/include/llassetgen/DistanceTransform.h:
```diff
@@ -1,73 +1,57 @@
-        std::unique_ptr<OutputType[]> output;
-        template <typename PixelType>
-        void exportPngInternal(png_struct_def*, OutputType, OutputType);
+        template <typename PixelType, bool flipped = false, bool invalidBounds = false>
```
I think the usual C++ convention is to write type and value template parameters the same way (CamelCase in our case). I would find it useful because it differentiates runtime and compile time values.

Felerius commented on source/llassetgen/include/llassetgen/DistanceTransform.h:
```diff
@@ -1,73 +1,57 @@
-
-        virtual void transform() = 0;
+        const Image& input;
+        const Image& output;
```
Do these need to be public? If the general usage is "create instance" -> "call transform" -> "do something with output" you have access to the images anyway, because you passed them in when creating the instance.

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        LLASSETGEN_NO_EXPORT static void flushData(png_struct_def* png);
+
+       public:
+        ~Image();
```
I don't know the rules for implicitly declared copy constructors/assignment operators by heart, but should they be declared we could easily end up doubly deleting `data` (create copy, delete copy, delete original). I'd think to be sure we should be explicit with regards to copy/move (following the rule of five):
```cpp
Image(const Image&) = delete;
Image(Image&&) = default;
Image& operator=(const Image&) = delete;
Image& operator=(Image&&) = default;
```

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        size_t stride;
+        uint8_t bitDepth;
+        uint8_t* data;
+        bool isOwnerOfData;
```
This is true exactly when the image is not a view, right? If so, I would name this `isView`, and note on the `view` method that views are only valid while their original image exists (which wasn't the case when we were using shared pointers).

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        Image view(Vec2<size_t> _min, Vec2<size_t> _max, size_t padding = 0);
+        size_t getWidth() const;
+        size_t getHeight() const;
+        size_t getBitDepth() const;
```
The usual naming convention is avoiding a `get` prefix. In the case of bit depth the variable also needs to be renamed, I used something like `bitDepth_` in the packer classes.

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        size_t getHeight() const;
+        size_t getBitDepth() const;
+        bool isValid(Vec2<size_t> pos) const;
+        template <typename pixelType>
```
For consistency this should be `PixelType`.

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        void clear() const;
+
+        void load(const FT_Bitmap_& ft_bitmap);
+        Image(const std::string& filepath, uint8_t _bitDepth = 0);
```
Should probably be grouped with the other constructors

Felerius commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,166 @@
+            return 0;
+        if (flipped)
+            std::swap(pos.x, pos.y);
+        return image.getPixel<PixelType>(pos);
```
This (and also `setPixel`) is very unclear. What would be the downside to having two `getPixel*` methods, both calling a free `getPixel<PixelType, Flipped, InvalidBounds>(pos, image)` function?

Also, replacing `flipped` and `invalidBounds` with two value enums could maybe make the calling code clearer (something like `getOutputPixel<Flipped, DontCheckBounds>` instead of `getOutputPixel<true, false>`)

Felerius commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,166 @@
+                                getPixel<InputType, false, true>({x, y - 1}) != center || getPixel<InputType, false, true>({x, y + 1}) != center))
+                    ? 0
+                    : std::numeric_limits<OutputType>::infinity()
+                );
```
Probably easier to read if the pixel values are read into variables first.

Felerius commented on source/llassetgen/source/DistanceTransform.cpp:
```diff
@@ -1,300 +1,166 @@
+                        square(begin == 0
+                            ? (end - i)                                            // Falling slope
+                            : ((i < (end + begin) / 2) ? i - begin + 1 : end - i)  // Rising and falling slope
+                        )
```
Mostly screwed up by the formatter, but even with hand formatting this could still be split into multiple lines for readability.

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        void setPixel(Vec2<size_t> pos, pixelType data) const;
+        template <typename pixelType = uint8_t>
+        void fillRect(Vec2<size_t> _min, Vec2<size_t> _max, pixelType in = 0) const;
+        void clear() const;
```
Why are these three methods const? While it is legal C++ because the data is behind a pointer, it is very misleading (`const std::vector` also doesn't allow editing values). Is there any specific reasons that these need to stay const?

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            byte <<= bit_pos * bitDepth;
+            byte >>= 8 - bitDepth;
+            uint32_t casted = static_cast<uint32_t>(byte);
+            return reinterpret_cast<pixelType&>(casted);
```
I find it way harder to follow whats going on here compared to the pointer version (especially because I've never seen `reinterpret_cast` used with a reference ever before).

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            uint8_t mask = 0xFF;
+            uint8_t in_byte = static_cast<uint8_t>(reinterpret_cast<uint32_t&>(in));
+            mask <<= 8 - bitDepth;
+            mask >>= 8 - bitDepth;
```
`uint8_t mask = ~(0xFF << bitDepth)` would be shorted and also easier to understand in my opinion.

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            memset(data, 0, stride * getHeight());
+        } else {
+            fillRect(min, max, 0);
+        }
```
Another reason to rename `isOwnerOfData` to `isView`, or at least document the relation somewhere.

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            assert(stride == static_cast<size_t>(ft_bitmap.pitch));
+            memcpy(data, ft_bitmap.buffer, ft_bitmap.pitch * ft_bitmap.rows);
+        } else if (min.x % 8 == 0) {
+            assert(bitDepth == 1);
```
Is there a reason that this always has to hold, or is this simply not implemented for bitDepth != 1?

Felerius commented on source/llassetgen/include/llassetgen/Image.h:
```diff
@@ -0,0 +1,52 @@
+        LLASSETGEN_NO_EXPORT Image(Vec2<size_t> _min, Vec2<size_t> _max, size_t _stride, uint8_t _bitDepth,
+                                   uint8_t* _data);
+        LLASSETGEN_NO_EXPORT static uint32_t reduceBitDepth(uint32_t in, uint8_t in_bitDepth, uint8_t out_bitDepth);
+        LLASSETGEN_NO_EXPORT static size_t getFtBitdepth(const FT_Bitmap_& ft_bitmap);
```
Should probably be `getFtBitDepth` for consistency (or just `ftBitDepth`).

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+
+    void Image::flushData(png_structp png) {
+        png_voidp a = png_get_io_ptr(png);
+        ((std::ostream*)a)->flush();
```
Should probably be `reinterpret_cast` for all three.

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+        if (!in_file.good()) {
+            std::cerr << "could not read from file";
+            abort();
+        }
```
We don't have a good error handling strategy for the library yet, but writing to stdout and aborting is definitely not it.

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            abort();
+        }
+
+        png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
```
`NULL` -> `nullptr`

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+
+        png_infop info = png_create_info_struct(png);
+        if (!info) {
+            png_destroy_read_struct(&png, (png_infopp)0, (png_infopp)0);
```
The latter two arguments could also be `nullptr` I think?

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            PNG_COLOR_TYPE_GRAY_ALPHA
+            PNG_COLOR_MASK_PALETTE
+            PNG_COLOR_MASK_COLOR
+            PNG_COLOR_MASK_ALPHA*/
```
Should probably be reported as an error somehow, instead of leaving the image in a uninitialized state.

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+        size_t png_stride = (getWidth() * png_bitDepth + 7) / 8;
+
+        std::unique_ptr<png_bytep> row_ptrs(new png_bytep[getHeight()]);
+        std::unique_ptr<uint8_t> multi_channel_data(new uint8_t[getHeight() * png_stride * channels]);
```
Couldn't this be `std::make_unique<png_bytep[]>(getHeight())` (and respectively for `multi_channel_data`)? Would also make assignment easier (`row_ptrs.get()[i]` -> `row_ptrs[i]`).

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+            png_set_swap(png);
+        }
+
+        // TODO
```
Whats left to do here (if anything)?

Felerius commented on source/llassetgen/source/Image.cpp:
```diff
@@ -0,0 +1,360 @@
+        png_write_end(png, nullptr);
+
+        png_destroy_write_struct(&png, &info);
+        out_file.close();
```
Streams are implicitly closed when they go out of scope (same goes for the importing constructor).

Felerius commented on source/tests/llassetgen-tests/Image.cpp:
```diff
@@ -0,0 +1,129 @@
+
+// current working directory: openll-asset-generator/build/source/tests
+std::string test_source_path = "../../../source/tests/llassetgen-tests/testfiles/";
+std::string test_destination_path = "../../";
```
This relies on the location of the cmake build dir compared to the source directory, making it hard to run tests while developing (especially with an IDE). Could we maybe pass the path to the source directory in using CMake (cmake variable -> c++ define -> here)?

In a perfect world we would also use the temp dir as the destination, or just clean up the files after the tests.

Felerius commented on source/tests/llassetgen-tests/Image.cpp:
```diff
@@ -0,0 +1,129 @@
+    EXPECT_EQ(pixel_size_set, 0);
+
+    FT_Error glyph_loaded = FT_Load_Glyph(face, FT_Get_Char_Index(face, letter), FT_LOAD_RENDER | FT_LOAD_MONOCHROME);
+    EXPECT_EQ(glyph_loaded, 0);
```
If the only thing the test is doing should the condition not be met is crashing then these should be `ASSERT_EQ`s

Felerius commented on source/tests/llassetgen-tests/Image.cpp:
```diff
@@ -0,0 +1,129 @@
+    Image glyph_view = canvas.view({0, 0}, {x, y}, padding);
+    glyph_view.load(face->glyph->bitmap);
+
+    Image invert_view = glyph_view.view(Vec2<size_t>(10, 10), Vec2<size_t>(20, 20));
```
These vectors (and many others in the tests) could be written as `{x, y}`.

Felerius commented on source/tests/llassetgen-tests/Image.cpp:
```diff
@@ -0,0 +1,129 @@
+    std::unique_ptr<DistanceTransform> dt(new DeadReckoning(input, output));
+    dt->transform();
+    output.exportPng<DistanceTransform::OutputType>(test_destination_path + "DeadReckoning.png", -20, 50);
+    EXPECT_EQ(1, 1);
```
Is this required by googletest?
